File: accbot-android/app/src/main/java/com/accbot/dca/data/local/Bip39WordList.kt
package com.accbot.dca.data.local

import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import java.security.MessageDigest
import java.security.SecureRandom
import javax.inject.Inject
import javax.inject.Singleton

/**
 * BIP39 wordlist loaded from assets + 12-word seed generator.
 *
 * Uses 128-bit entropy → SHA-256 checksum (4 bits) → 132 bits → 12 × 11-bit indices.
 */
@Singleton
class Bip39WordList @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val words: List<String> by lazy {
        context.assets.open("bip39_english.txt").bufferedReader().useLines { lines ->
            lines.filter { it.isNotBlank() }.toList()
        }
    }

    fun generateSeed(): List<String> {
        val entropy = ByteArray(16) // 128 bits
        SecureRandom().nextBytes(entropy)
        return entropyToWords(entropy)
    }

    fun isValidWord(word: String): Boolean =
        words.contains(word.lowercase())

    fun isValidSeed(seedWords: List<String>): Boolean {
        if (seedWords.size != 12) return false
        if (seedWords.any { !isValidWord(it) }) return false

        // Verify checksum: convert words back to bits, check last 4 bits match SHA-256
        val indices = seedWords.map { w -> words.indexOf(w.lowercase()) }
        if (indices.any { it < 0 }) return false

        // 12 words × 11 bits = 132 bits = 128 entropy + 4 checksum
        val bits = BooleanArray(132)
        for (i in indices.indices) {
            val idx = indices[i]
            for (b in 0 until 11) {
                bits[i * 11 + b] = (idx shr (10 - b)) and 1 == 1
            }
        }

        val entropyBytes = ByteArray(16)
        for (i in 0 until 128) {
            if (bits[i]) entropyBytes[i / 8] = (entropyBytes[i / 8].toInt() or (1 shl (7 - i % 8))).toByte()
        }

        val hash = MessageDigest.getInstance("SHA-256").digest(entropyBytes)
        val checksumBits = (hash[0].toInt() and 0xFF) shr 4 // top 4 bits
        var actualChecksum = 0
        for (b in 0 until 4) {
            if (bits[128 + b]) actualChecksum = actualChecksum or (1 shl (3 - b))
        }

        return checksumBits == actualChecksum
    }

    fun getSuggestions(prefix: String, limit: Int = 5): List<String> {
        if (prefix.isBlank()) return emptyList()
        val lower = prefix.lowercase()
        return words.filter { it.startsWith(lower) }.take(limit)
    }

    internal fun entropyToWords(entropy: ByteArray): List<String> {
        require(entropy.size == 16) { "Entropy must be 16 bytes (128 bits)" }

        val hash = MessageDigest.getInstance("SHA-256").digest(entropy)
        // checksum = top 4 bits of SHA-256(entropy)
        val checksumByte = hash[0]

        // 128 entropy bits + 4 checksum bits = 132 bits
        val bits = BooleanArray(132)
        for (i in 0 until 128) {
            bits[i] = (entropy[i / 8].toInt() shr (7 - i % 8)) and 1 == 1
        }
        for (b in 0 until 4) {
            bits[128 + b] = (checksumByte.toInt() shr (7 - b)) and 1 == 1
        }

        // 132 bits / 11 = 12 words
        return List(12) { wordIndex ->
            var idx = 0
            for (b in 0 until 11) {
                if (bits[wordIndex * 11 + b]) idx = idx or (1 shl (10 - b))
            }
            words[idx]
        }
    }
}

File: accbot-ios/AccBot/Data/Local/Bip39WordList.swift
import Foundation
import CommonCrypto

/// BIP39 wordlist loaded from bundle + 12-word seed generator.
///
/// Uses 128-bit entropy -> SHA-256 checksum (4 bits) -> 132 bits -> 12 x 11-bit indices.
final class Bip39WordList {
    private lazy var words: [String] = {
        guard let url = Bundle.main.url(forResource: "bip39_english", withExtension: "txt"),
              let content = try? String(contentsOf: url, encoding: .utf8) else {
            return []
        }
        return content.components(separatedBy: .newlines).filter { !$0.isEmpty }
    }()

    func generateSeed() -> [String] {
        var entropy = Data(count: 16) // 128 bits
        entropy.withUnsafeMutableBytes { _ = SecRandomCopyBytes(kSecRandomDefault, 16, $0.baseAddress!) }
        return entropyToWords(entropy)
    }

    func isValidWord(_ word: String) -> Bool {
        words.contains(word.lowercased())
    }

    func isValidSeed(_ seedWords: [String]) -> Bool {
        guard seedWords.count == 12 else { return false }
        guard seedWords.allSatisfy({ isValidWord($0) }) else { return false }

        let indices = seedWords.map { w in words.firstIndex(of: w.lowercased()) ?? -1 }
        guard indices.allSatisfy({ $0 >= 0 }) else { return false }

        // 12 words x 11 bits = 132 bits = 128 entropy + 4 checksum
        var bits = [Bool](repeating: false, count: 132)
        for (i, idx) in indices.enumerated() {
            for b in 0..<11 {
                bits[i * 11 + b] = (idx >> (10 - b)) & 1 == 1
            }
        }

        var entropyBytes = Data(count: 16)
        for i in 0..<128 {
            if bits[i] {
                entropyBytes[i / 8] |= UInt8(1 << (7 - i % 8))
            }
        }

        var hash = Data(count: Int(CC_SHA256_DIGEST_LENGTH))
        entropyBytes.withUnsafeBytes { entropyPtr in
            hash.withUnsafeMutableBytes { hashPtr in
                _ = CC_SHA256(entropyPtr.baseAddress, CC_LONG(entropyBytes.count), hashPtr.baseAddress?.assumingMemoryBound(to: UInt8.self))
            }
        }

        let checksumBits = Int(hash[0]) >> 4 // top 4 bits
        var actualChecksum = 0
        for b in 0..<4 {
            if bits[128 + b] { actualChecksum |= 1 << (3 - b) }
        }

        return checksumBits == actualChecksum
    }

    func getSuggestions(prefix: String, limit: Int = 5) -> [String] {
        guard !prefix.isEmpty else { return [] }
        let lower = prefix.lowercased()
        return Array(words.filter { $0.hasPrefix(lower) }.prefix(limit))
    }

    func entropyToWords(_ entropy: Data) -> [String] {
        precondition(entropy.count == 16, "Entropy must be 16 bytes (128 bits)")

        var hash = Data(count: Int(CC_SHA256_DIGEST_LENGTH))
        entropy.withUnsafeBytes { entropyPtr in
            hash.withUnsafeMutableBytes { hashPtr in
                _ = CC_SHA256(entropyPtr.baseAddress, CC_LONG(entropy.count), hashPtr.baseAddress?.assumingMemoryBound(to: UInt8.self))
            }
        }
        let checksumByte = hash[0]

        // 128 entropy bits + 4 checksum bits = 132 bits
        var bits = [Bool](repeating: false, count: 132)
        for i in 0..<128 {
            bits[i] = (Int(entropy[i / 8]) >> (7 - i % 8)) & 1 == 1
        }
        for b in 0..<4 {
            bits[128 + b] = (Int(checksumByte) >> (7 - b)) & 1 == 1
        }

        // 132 bits / 11 = 12 words
        return (0..<12).map { wordIndex in
            var idx = 0
            for b in 0..<11 {
                if bits[wordIndex * 11 + b] { idx |= 1 << (10 - b) }
            }
            return words[idx]
        }
    }
}
